Firestore/core/src/firebase/firestore/api/write_batch.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_WRITE_BATCH_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_API_WRITE_BATCH_H_
+
+#if !defined(__OBJC__)
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#import "Firestore/Source/Core/FSTFirestoreClient.h"
+#import "Firestore/Source/Model/FSTMutation.h"
+
+#include "Firestore/core/src/firebase/firestore/api/document_reference.h"
+#include "Firestore/core/src/firebase/firestore/api/firestore.h"
+#include "Firestore/core/src/firebase/firestore/api/input_validation.h"
+#include "Firestore/core/src/firebase/firestore/core/user_data.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class FSTMutation;
+
+namespace firebase {
+namespace firestore {
+namespace api {
+
+class WriteBatch {
+ public:
+  WriteBatch() = default;
+  explicit WriteBatch(std::shared_ptr<Firestore> firestore)
+      : firestore_{firestore} {
+  }
+
+  void SetData(const DocumentReference &reference,
+               core::ParsedSetData &&setData) {
+    VerifyNotCommitted();
```

Comment:
Nit: please create a source file and most of the member function definitions there (I'd move everything, or everything except the constructor).

Off the top of my head, I can think of two reasons, readability and compile times:
* when definitions are separate, looking at a declaration is a very convenient way to see the class interface without having to scroll through implementation details. It's something I miss in other languages;
* if you change a function definition in the source file, only the source file needs to be recompiled. If, however, the function definition is in the header file, everything including that header will have to be recompiled;
* it allows moving some `include` directives to the source file, once again potentially reducing the need to recompile.

Reply:
(tangentially related note: with a modern compiler, you can expect it to basically ignore any `inline` declarations as far as optimization is concerned. In this case, inlining or not is purely a stylistic choice)